### PR TITLE
Password validation

### PR DESCRIPTION
**Pull request by ninjakx**
[#234](https://github.com/systers/portal/issues/234/)

portal1
portal2
portal3

### CODE REVIEW

File: systers_portal/users/adapter.py [hunk 1]
```diff
@@ -1,11 +1,60 @@
...
+        if len(password) >= 6:
+           
+            if re.match(r'^(?=.*?[0-9])(?=.*[A-Z])', password) and set('[~!@#$%^&*()_+{}":;\']+$').intersection(password):
+                return password
+
+            elif not re.match(r'^(?=.*?[0-9])', password) and  not re.match(r'^(?=.*?[A-Z])', password) and set('[~!@#$%^&*()_+{}":;\']+$').intersection(password):
+                raise ValidationError("Password must have at least one uppercase letter and one digit")
+
+            elif not  set('[~!@#$%^&*()_+{}":;\']+$').intersection(password) and not re.match(r'^(?=.*?[0-9])', password) and re.match(r'^(?=.*[A-Z])', password):
+                raise ValidationError("Password must have at least one Special character and one digit")
```
Comment (tapasweni-pathak) [inline systers_portal/users/adapter.py]: r/S/s in special.

File: systers_portal/users/adapter.py [hunk 2]
```diff
@@ -1,11 +1,60 @@
...
+
+            elif not  set('[~!@#$%^&*()_+{}":;\']+$').intersection(password) and not re.match(r'^(?=.*[A-Z])', password) and re.match(r'^(?=.*?[0-9])', password):
+                raise ValidationError("Password must have at least one Uppercase letter and one Special character")
```
Comment (tapasweni-pathak) [inline systers_portal/users/adapter.py]: r/U/u and r/S/s.

File: systers_portal/users/adapter.py [hunk 3]
```diff
@@ -1,11 +1,60 @@
...
+                raise ValidationError("Password must have at least one special character")
+
+            elif not re.match(r'^(?=.*[A-Z])', password) and re.match(r'^(?=.*?[0-9])', password) and set('[~!@#$%^&*()_+{}":;\']+$').intersection(password):
+                raise ValidationError("Password must have at least one Uppercase letter")
+
+            elif not re.match(r'^(?=.*?[0-9])', password) and set('[~!@#$%^&*()_+{}":;\']+$').intersection(password) and re.match(r'^(?=.*[A-Z])', password):
+                raise ValidationError("Password must have at least one digit")
+
+            else :
+                raise ValidationError("Password must have at least one Uppercase letter and one Special character and one digit")
```
Comment (tapasweni-pathak) [inline systers_portal/users/adapter.py]: r/U/u and r/S/s.

File: systers_portal/users/adapter.py [hunk 4]
```diff
@@ -1,11 +1,60 @@
...
+
+        
     def get_login_redirect_url(self, request):
         return reverse('user', args=[request.user.username])
 
     def get_signup_redirect_url(self, request):
         return reverse('user', args=[request.user.username])
+
+
+
```
Comment (tapasweni-pathak) [inline systers_portal/users/adapter.py]: Have you flake8ed this?

File: systers_portal/users/adapter.py (hunk 1 above)
Comment (ninjakx) [reply systers_portal/users/adapter.py]: r/S/s ????

**Comment by ninjakx** [comment]
@tapasweni-pathak - I am not able to squash all the commits into a single commit .
also i am not getting you . I don't know how to use git other than making repo and pushing it.
How should i clean the commits ?

**Comment by tapasweni-pathak** [comment]
> I am not able to squash all the commits into a single commit .
also i am not getting you . I don't know how to use git other than making repo and pushing it.
How should i clean the commits ?

Above with r/S/s e.t.c., I mean replace the capital S in the word *Special* with *special*, same for others.

For Squashing:
1. Push more commits on this pull request for above comments. 
2. Say ```git rebase -i <hash>``` the first commit in these commit message.
3. Replace all but first ```pick``` with ```s``` or ```squash```. Save the file.
4. Now provide a proper commit message for all your changes. 

This will combine your commits and diffs into one. If you face any issues, let me know!

**Comment by ninjakx** [comment]
@tapasweni-pathak - I am getting noop  when i do git rebase -i

**Comment by tapasweni-pathak** [comment]
> Is it necessary to pass this check? I have to change the whole error messages in order to pass the travis check :(.
I have to change this to "Password must have atleast one digit ,one uppercase and one special character" to "Missing : 1 digit,1 uppercase & 1 special character"

No, don't change the error message text, bring part of it to newline. Did you run flake8 on your changes?

File: systers_portal/users/adapter.py [hunk 5]
```diff
@@ -1,9 +1,61 @@
...
+    def clean_password(self, password):
+        # Password should have at least one uppercase letter , one digit and one special character
+        x = r'^(?=.*?[0-9])(?=.*[A-Z])'
+        y = '[~!@#$%^&*()_+{}":;\']+$'
+        z = r'^(?=.*?[0-9])'
+        u = r'^(?=.*?[A-Z])'
+        a = re.match(z, password)
+        b = re.match(x, password)
+        c = set(y).intersection(password)
+        d = re.match(u, password)
```
Comment (tapasweni-pathak) [inline systers_portal/users/adapter.py]: [There is a need to rename these variables](https://www.python.org/dev/peps/pep-0008/#id37).

**Pull request merged**